R/visualize.R

```diff
@@ -0,0 +1,237 @@
+                         fairness_data[, fairness_name, with=FALSE]),
+              id.vars = "model")
+  ggplot(data, aes(x=model, y=value, fill=model)) +
+    geom_bar(stat = "identity",  width=1/length(unique(data$model))) +
+    xlab("metrics") +
+    ylab("values") +
+    theme(legend.position = "none") +
+    scale_fill_hue(c=100, l=60) +
+    facet_wrap(~variable)
+}
+
+#' @export
+fairness_compare.ResampleResult <- function(object, fairness_measure = msr("fairness.acc"), ...){
+  object = as_benchmark_result(object)
+  fairness_compare(object, fairness_measure)
+}
+
+#' Fairness Prediction Density Visualization
+#'
+#' @description
+#' This function specializes in visualizing the plots that will demonstrate the predicted probability density for each level in the protected column.
+#' Which will potentially reveal the fairness problems in the dataset if the density differs significantly for each level.
+#' The plot is a combination of boxplot and violin plot. So users could use the plot to detect both the outliers and distribution problems.
+#' The y-axis shows the levels in protected columns. And the x-axis shows the predicted probability.
+#' The title for the plot will demonstrate which class for predicted probability.
+#'
+#' @param object ([PredictionClassif])\cr ([ResampleResult])\cr ([BenchmarkResult])\cr
+#' The binary class prediction object that will be evaluated. Only one data task is allowed for Resample Result and Benchmark Result.
+#'
+#' @param ...
+#' The arguments to be passed to methods, such as:
+#'
+#' * `task` ([TaskClassif])\cr
+#' The data task that contains the protected column.
+#'
+#' @examples
+#' library(mlr3fairness)
+#' library(mlr3learners)
+#' library(mlr3)
+#' library(ggplot2)
+#' library(data.table)
+#'
+#' task = tsk("adult_train")
+#' learner = lrn("classif.ranger", predict_type = "prob")
+#' learner$train(task)
+#' predictions = learner$predict(task)
+#' fairness_prediction_density(predictions, task)
+#' @export
+fairness_prediction_density <- function(object, ...){
+  UseMethod("fairness_prediction_density")
+}
+
+#' @export
+fairness_prediction_density.PredictionClassif<- function(object, task, ...){
+  data <- melt(data.table(task$data(cols = task$col_roles$pta),
+                          object$prob[,1]),
+               id = task$col_roles$pta)
+  colnames(data) <- c("protected_variable", "variable", "probability")
+
+  ggplot(data, aes(x = protected_variable, y=probability)) +
+    geom_boxplot(aes(fill = protected_variable), width=0.4/length(unique(data$protected_variable))) +
+    geom_violin(alpha = 0.3, width=1.5/length(unique(data$protected_variable)), fill = "grey") +
+    xlab("Protected attributes") +
+    ylab( paste0("predicted probability for ", colnames(object$prob)[1])) +
+    theme(legend.position = "none") +
+    scale_fill_hue(c=100, l=100) +
+    ylim(c(0,1)) +
+    coord_flip()
+}
+
+#' @export
+fairness_prediction_density.BenchmarkResult <- function(object, task, ...){
```

Comment:
This can do the same thing as above via facet_wrap by learner?

Reply:
Yes. The function for BenchmarkResult basically did the same thing.